File: j2kt/jre/java/native/java/util/concurrent/atomic/AtomicLong.java
// CHECKSTYLE_OFF: Copyrighted to Guava Authors.
/*
 * Copyright (C) 2011 The Guava Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// CHECKSTYLE_ON

package java.util.concurrent.atomic;

import javaemul.internal.annotations.KtName;
import javaemul.internal.annotations.KtNative;
import org.jspecify.nullness.NullMarked;

/** JRE AtomicLong stub; see AtomicLong.kt for the actual wrapper. */
@KtNative(name = "java.util.concurrent.atomic.AtomicLong")
@NullMarked
public class AtomicLong extends Number implements java.io.Serializable {

  public AtomicLong(long initialValue) {}

  public AtomicLong() {}

  public native long get();

  public native void set(long newValue);

  public native void lazySet(long newValue);

  public native long getAndSet(long newValue);

  public native boolean compareAndSet(long expect, long update);

  public native long getAndIncrement();

  public native long getAndDecrement();

  public native long getAndAdd(long delta);

  public native long incrementAndGet();

  public native long decrementAndGet();

  public native long addAndGet(long delta);

  @Override
  public native String toString();

  @Override
  @KtName("toInt")
  public native int intValue();

  @Override
  @KtName("toLong")
  public native long longValue();

  @Override
  @KtName("toFloat")
  public native float floatValue();

  @Override
  @KtName("toDouble")
  public native double doubleValue();
}

File: j2kt/jre/java/native/java/util/concurrent/atomic/AtomicLong.kt
/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package java.util.concurrent.atomic

@OptIn(kotlin.concurrent.atomics.ExperimentalAtomicApi::class)
@Suppress("DEPRECATION")
class AtomicLong
internal constructor(private val ktAtomicLong: kotlin.concurrent.atomics.AtomicLong) :
  kotlin.Number() {

  constructor() : this(kotlin.concurrent.atomics.AtomicLong(0L))

  constructor(initialValue: Long) : this(kotlin.concurrent.atomics.AtomicLong(initialValue))

  fun get(): Long = ktAtomicLong.load()

  fun set(newValue: Long) {
    ktAtomicLong.store(newValue)
  }

  fun lazySet(newValue: Long) {
    ktAtomicLong.store(newValue)
  }

  fun getAndSet(newValue: Long): Long = ktAtomicLong.exchange(newValue)

  fun compareAndSet(expect: Long, update: Long): Boolean =
    ktAtomicLong.compareAndSet(expect, update)

  fun getAndIncrement(): Long = ktAtomicLong.fetchAndAdd(1)

  fun getAndDecrement(): Long = ktAtomicLong.fetchAndAdd(-1)

  fun getAndAdd(delta: Long): Long = ktAtomicLong.fetchAndAdd(delta)

  fun incrementAndGet(): Long = ktAtomicLong.addAndFetch(1)

  fun decrementAndGet(): Long = ktAtomicLong.addAndFetch(-1)

  fun addAndGet(delta: Long): Long = ktAtomicLong.addAndFetch(delta)

  override fun toString(): String = ktAtomicLong.toString()

  override fun toInt(): Int = ktAtomicLong.load().toInt()

  override fun toLong(): Long = ktAtomicLong.load()

  override fun toFloat(): Float = ktAtomicLong.load().toFloat()

  override fun toDouble(): Double = ktAtomicLong.load().toDouble()

  override fun toByte(): Byte = ktAtomicLong.load().toByte()

  override fun toChar(): Char = ktAtomicLong.load().toChar()

  override fun toShort(): Short = ktAtomicLong.load().toShort()
}
